%\VignetteIndexEntry{Detection of de novo copy number alterations in case-parent trios}
%\VignetteDepends{?}
%\VignetteKeywords{MinimumDistance, copy number, SNP, case-parent trios, de novo}
%\VignettePackage{MinimumDistance}
\documentclass{article}
\usepackage{graphicx}
\usepackage{natbib}
\usepackage{url}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rcode}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textsf{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}
\newcommand{\R}{\textsf{R}}
\newcommand{\md}{\Rpackage{MinimumDistance}}
\newcommand{\md}{\Rpackage{MinimumDistance}}
\newcommand{\logRratio}{$\log_2$ R ratio}
\newcommand{\baf}{B allele frequency}
\newcommand{\bafs}{B allele frequencies}
\usepackage[margin=1in]{geometry}


\begin{document}
\title{Detection of de novo copy number alterations in case-parent
  trios using the \R{} package \md{}}
\date{\today}

\author{Moiz Bootwalla and Rob Scharpf}
\maketitle


<<setup, echo=FALSE, results=hide>>=
options(prompt="R> ", continue=" ", device=pdf, width=68)
@

\begin{abstract}

  This vignette describes the \ldots.

\end{abstract}

\section{Exemplar dataset}

This section should describe what data is required for the package and
how to format the data appropriately for reproducing our analysis.
{\it Ideally, we would include processed data in the package so that
  users can reproduce all of the analytic steps after preprocessing
  (e.g., after the BeadStudio analysis). }

This package requires
\begin{itemize}
  \item \logRratio{}s and \bafs{}.  How these files should be organized.
  \item chromosome and physical position for each marker
  \item a pedigree file (need an example of such a file)
  \item 'samplesheet' if available: contains info on gender, DNA
    source, chemistry plate, etc.
  \item \ldots
\end{itemize}

Here's a 3G HapMap file from CIDR that was processed using
BeadStudio.  This should contain BAFs and log R Ratios for roughly 30
HapMap CEU trios:

<<exampleFile>>=
example.beadStudioFile <- list.files("/thumper/ctsa/snpmicroarray/hapmap/processed/illumina/610kQuad",
				     pattern="CEU_Final",
				     full.names=TRUE)
c("character", "character", rep(NULL, 7), rep("numeric", 2))
header <- read.csv(example.beadStudioFile, skip=9,
		     colClasses=c("character", "character", rep("NULL", 7), rep("numeric", 2)))
trace(MinimumDistance:::initializeTrioContainer,browser)
MinimumDistance::::initializeTrioContainer(pedigree,samplesheet)
untrace(MinimumDistance:::initializeTrioContainer)
@

The sample sheet is in the same directory 'HumanHap610Quadv1_Sample_Map.csv'.

We really only need one father, mother, offspring trio for the
vignette, but it would be good to find an example of a trio with a de
novo amplification or deletion.  If not too big, we might include the
data for a few of the trios and a subset of the chromosomes (e.g., 5
trios and 4 chromosomes).  Example of the format currently required -- see

``/thumper/ctsa/beaty/holger/txtfiles''

You'll need the physical position and chromosome info.  The annotation
packages should be available from BioConductor. Here's how you can see
what annotation packages are available.  (You'll want to include this
in the vignette since other people may have different platforms).  The
R function minimumDistance initializes an object of class
'TrioSetList' and automatically provides annotation of chromosome and
physical position provided that the 'cdfname' for the package is
provided and an annotation package is available.

<<>>=
library(oligoClasses)
all.pkgs <- annotationPackages()
@

Show how to construct the container we use for analyzing trios from
the above files.

\begin{itemize}

  \item create matrix of B allele frequencies

  \item create matrix of \logRratio{}s

  \item create matrix of \bafs{}

  \item instantiate an object of class \Rclass{TrioSetList} object

  \end{itemize}

<<eval=FALSE>>=
container <- minimumDistance(path, ## for parsed data
			     samplesheet=samplesheet2,
			     pedigree=pedigree,
			     container.filename=container.filename,
			     chromosomes=chromosomes,
			     file.ext=".txt",
			     readFiles=readFiles,
			     calculate.mad=calculate.mad,
			     calculate.md=calculate.md,
			     exclusionRule=BeatyExperimentData:::excludeWGA,
			     cdfName="human610quadv1b")
@

<<>>=
load("~mbootwal/packages/MinimumDistance/data/trioSet-hapmap.rda")
##
library(MinimumDistance)
trace(minimumDistance, browser)
trioSetList <- minimumDistance(container=trioSetList, readFiles=FALSE, calculate.md=TRUE,
		       container.filename="~/trioSetList.rda")
mads <- matrix(NA, ncol(trioSetList[[1]]), 3)
colnames(mads) <- c("F", "M", "O")
rownames(mads) <- sampleNames(trioSetList[[1]])
## need to calculate mad
fff <- function(object, trio.index){
	lR <- logR(object)[, trio.index, ]
}
for(j in seq(length=length(rownames(mads)))){
	logrs <- do.call("rbind", sapply(trioSetList, fff, trio.index=j))
	mads[j, ] <- apply(logrs, 2, mad, na.rm=TRUE)
}
mad(trioSetList[[1]]) <- mads
##trace(MinimumDistance:::computeLoglik, browser)
unlink("~/results.rda")
rd <- minimumDistanceCalls(container=trioSetList, chromosomes=1, cbs.filename="~/results.rda")
@


\section{Detection of de novo copy number}

\subsection{Segmentation}

The function call looks something like this.

<<minimumDistanceCalls>>=
 ## segmentation
 ## classification
@


\subsection{Posterior probabilities of trio copy number states}

\subsection{Visualizations}

Plot the MAD log R ratio by chemistry plate

Plot the MAD log R ratio by DNA source

Plot the low-level summaries along with the minimum distance. Right
now, the function is called \Rfunction{minimumDistancePlot}.

'Rectangle' plots as in the paper (here, we'd only have de novo
predictions from 5 offspring)



\section{Session information}
<<sessionInfo, results=tex>>=
toLatex(sessionInfo())
@

\end{document}
