\documentclass{article}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage[authoryear,round,numbers]{natbib}
\usepackage{hyperref}
\usepackage[mathcal]{euscript}
\usepackage[margin=1in]{geometry}
\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}
\newcommand{\Rpackage}[1]{\textit{#1}}
\newcommand{\Rfunction}[1]{\texttt{#1}}
\newcommand{\Robject}[1]{\texttt{#1}}
\newcommand{\R}{\textsf{R}}
\newcommand{\bioc}{http://www.bioconductor.org}
\parindent 0in  % Left justify
\parskip 7.2pt
\newcommand{\denovo}{de novo}
\newcommand{\Denovo}{De novo}
\newcommand{\via}{\textit{via}}
\newcommand{\apriori}{\textit{a priorir}}
\newcommand{\adhoc}{\textit{ad hoc}}
\newcommand{\insertfigure}[1]{\vspace{1em}[FIGURE #1 ABOUT HERE]\vspace{1em}}
%\title{Denovo copy number variants implicated in cleft lip and palate
%  birth defects}
\title{Fast detection of \denovo{} copy number variants from
  genotyping arrays in a case-parent study of facial clefting}
%\title{Fast detection of \denovo{} copy number variants in case-parent
%trios}

\author{RS, TB, AS, HS, and IR (in some order)}

\begin{document}
\maketitle

\begin{abstract}

  Cleft palate and lip birth defects are heritable disorders.
  Previous genome wide association studies have identified several
  loci that are associated with facial clefting. Included among these
  studies is a large case-parent trio design for which high-throughput
  genotyping arrays were used to genotype approximately 7280
  individuals, or approximately 2400 trios.  While studies of
  transmission disequilibrium are on-going, it is not well understood
  whether the previously identified loci lie in regions of copy number
  variants or the extent of copy number dosage sensitive genes that
  contribute to clefting.  To evaluate the impact of copy number
  variants on facial clefting, we re-analyzed the Illumina genotyping
  arrays for \denovo{} copy number variants.  Our analyses reveal
  several offspring with hemizygous \denovo{} deletions in a region of
  chromosome 22 previously implicated in DiGeorge syndrome.  These
  regions were validated \via{} multiplex qPCR.  Our approach also
  identified regions on chromosomes \ldots that have not been
  previously linked to cleft palate but show a high frequency of
  \denovo{} alterations in the present study.  To identify these
  hotspots of \denovo{} recombination, we describe a simple and fast
  approach for \denovo CNV detection as an alternative to a hidden
  Markov model. Our approach appears to be more robust to noisy
  samples and genomic waves. We combine visualizations of inferred
  segmental alterations with functional annotation available from
  existing \R{} packages to focus downstream analyses.

\end{abstract}

\section{Introduction}
\label{sec:intro}

Cleft palate and cleft lip affect 1 in every 2500 offspring.

%This is the first paper to identify copy number variants implicated in
%clefting (?)

Barber2008 JC describe a duplication in 8p23.1 may be associated with
a facial phenotype

Shi2009 microdeletions in candidate genes for cleft lip / cleft
palate.  candidate gene analysis.1200 snps, 333 candidate genes.  We
confirmed deletions involving CYP1B1, FGF10, SP8, SUMO1, TBX1, TFAP2A,
and UGT7A1, including both de novo and familial cases. Deletions of
SUMO1, TBX1, and TFAP2A are likely to be etiologic.


duplications and deletions in digeorge region on chr22 (Sivertsen2007,


cnvs that phenocopy the chromosome 22q11.2 deletion syndrome
(Busse2011)
% 22q11 Deletion Syndrome includes the overlapping phenotypes of
% DiGeorge/Velocardiofacial Syndromes, characterized by conotruncal
% heart defects, cleft palate, thymus, and parathyroid gland

Letra2010 assess association between nonsyndromic (NS) cleft lip with or without cleft palate (CL(P)) and SNPs within the CRISPLD2 gene


Jugessur2009  357 candidate genes

Birnbaum2009, Grant2009 GWAS for NSCL(P) on 8q24

Mossey2009 Review article in lancet


\section{Methods}
\label{sec:methods}

BeadStudio was used to process the raw intensities for 7599 samples,
including 318 HapMap samples that were processed simultaneously by the
CIDR genotyping core. Quality control of samples for CNV analysis was
assessed primarily through the median absolute deviation (MAD) of the
log R ratios estimated for each sample across chromosomes 1 through
22.  We stratified variance estimates of the log R ratio by plate and
DNA source to assess the impact of experimental and technological
sources of variance that can give rise to batch effects
\citep{Leek2010}.

% denovo is not modeled as a hidden state.
%
Regions of \denovo{} CNV were inferred by two approaches. The first
approach we applied was a HMM for case-parent trios implemented in the
PennCNV software, referred to as the joint HMM.  The joint HMM
incorporates signal intensities from the log R ratio ($r$) and the B
allele frequency ($b$) from the father-mother-offspring trio.  An
indicator variable for the \denovo{} event status of the offspring
allows for non-Mendelian patterns of inheritance. Conditional on the
underlying hidden states $z$, the likelihood for the distribution of
the trios can be factored as a mixture of a multivariate normal
distribution for the $r's$ and a truncated multivariate normal for the
$b's$.

transition probs of CN states: The transition probabilities for the
parents are modeled as a function of the distance (includes a gamma
term that is unknown. What is the gamma for and how is it estimated?).
The transition probability for the offspring includes a term

(model mendelian inheritance using models for chromosome-specific copy
numbers given the total copy number at 2 adjancent SNPs.  use a single
parameter b that specificies the prob. of the less likely
chromosome-specific CN configuration. 25 x 25 x 25 table for 2-marker
CN inheritance.)

transition probs of denovo event for offspring: use same tp as used
for CN states, but with different transition parameters (delta instead
of gamma).  Gamma and Delta estimation:
Fix a and b at 0.0009. denovo rate epsilon and Delta estimated from
previously published HapMap CNV results.  Baum-Welch algorithm is used
to estimate Gamma. Viterbi algorithm used to find optimal state path.

The joint HMM distribution of
log R ratios and B allele frequencies. The hidden states of this HMM
are \ldots.

We also explored an alternative to summarizing the joint distribution
of log R ratios.  At each marker, we subtract the estimated log R
ratio of the offspring from the offspring's mother and father.  Both
minimum and maximum distances are retained.  To identify \denovo{}
deletions, we smoothed the minimum distance as a function of physical
position using circular binary segmentation (CBS; \cite{}). Parameter
settings used for CBS are provided in Section \ref{sec:sm}.  We use an
\adhoc{} approach for calling \denovo{} deletions that is a function
of the amplitidue of the minimum distance and coverage. Figure X plots
the minimum segment mean required for calling a deletion versus
coverage. Implicitly, the uncertainty of the segment mean is larger
for regions with less coverage and less for regions with high
coverage.

\insertfigure{1}

The procedure for inferring \denovo{} amplifications is analogous,
with a segmentation of the maximum difference between offspring and
parents followed by CBS and calling the segment means as \denovo{}
using the same procedure as illustrated in Figure 1.



\section{Results}

\paragraph{Obtaining a set of high-quality samples for CNV analysis}
The noise of BeadStudio log R ratios was strongly influenced by WGA,
and this affected subsequent inference via a HMM.  Trio were excluded
from subsequent analyses if any member in the trio had DNA that was
WGA or if the MAD of the log R ratios exceeded 0.3.  We also assessed
whether the number of \denovo{} events was associated with
site. Supplementary Figure X shows that samples from site 23 tended to
have higher \denovo{} call rates than other sites.  Further analysis
revealed that the high call rates were primarily on two of the plates
from site 23. Did the samples on these plates have abnormally high log
R ratios?

\begin{figure}[t!]
  \mbox{
  \includegraphics[width=0.48\textwidth]{figures/wgaEffect_01.pdf}
  \includegraphics[width=0.48\textwidth]{figures/wgaEffect_02.pdf}
}
\caption{Impact of DNA source on noise of the BeadStudio log R ratios
  (left) and downstream inference of \denovo{} events using PennCNV
  (right).}
\end{figure}

\paragraph{High frequency \denovo{} CNV} The joint HMM provides a
likelihood based approach for inferring the genomic start and stop
coordinates of the \denovo{} events.  Figure X plots genomic position
on the x-axis versus the offspring index.  The shaded rectanges
indicate inferred \denovo{} events and the color codes indicate
whether the \denovo{} event corresponds to a copy number deletion or
duplication in the offspring. Vertical stripes in this figure
correspond to sites for which there appears to be a frequent
recombination event from parent to offspring that results in loss or
gain of DNA copy number in the offspring.  There are also many
horizontal stripes.  Horizontal stripes likely represent technological
artifact.  (TODO: Ordering the offsrping by (site, dna source, MAD)
shows that the horizontal stripes tend to occur for \dots.  This
regions can also be a result of genomic waves that perhaps effect one
individual in the trio more than the others.  Genomic waves typically
occur as a result of different hybridization intensities due to
sequence content of the probes.  A wave correction prior to fitting
the HMM may reduce the susceptiblity of \denovo{} inferences to
waves.)  Closer inspection of many of the inferred alterations were
likely false positives.  (todo: explain why the false positives may be
occurring in the hmm).

In order to reduce the occurence of horizontal stripes and false
positive signals that tend to obsure the sites of recurrent \denovo{}
CNV, we implemented a simple alternative that directly estimates the
smallest distance between the case-parent log R ratios in order to
find the most similar profiles.  Figure X plots the inferred \denovo{}
events for the segmentation.



\section{Discussion}


\section{Supplemental Materials}
\label{sec:sm}

\end{document}

\section{Session Info}

<<sessionInfo, echo=FALSE, results=tex>>=
toLatex(sessionInfo())
@

