\documentclass{article}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage[authoryear,round,numbers]{natbib}
\usepackage{hyperref}
\usepackage[mathcal]{euscript}
\usepackage[margin=1in]{geometry}
\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}
\newcommand{\Rpackage}[1]{\textit{#1}}
\newcommand{\Rfunction}[1]{\texttt{#1}}
\newcommand{\Robject}[1]{\texttt{#1}}
\newcommand{\R}{\textsf{R}}
\newcommand{\bioc}{http://www.bioconductor.org}
\parindent 0in  % Left justify
\parskip 7.2pt
\newcommand{\denovo}{de novo}
\newcommand{\Denovo}{De novo}
\newcommand{\via}{\textit{via}}

\title{Denovo copy number variants implicated in cleft lip and palate
  birth defects}

\author{RS, TB, AS, HS, and IR (in some order)}

\begin{document}
\maketitle

\begin{abstract}

  Cleft palate and lip birth defects are heritable disorders.
  Previous genome wide association studies have identified several
  loci that are associated with the cleft phenotype.  We analyzed a
  collection of 2600 case-parent trios for \denovo{} copy number
  variants using the Illumina high-throughput genotyping platform. We
  identified several offspring with similar \denovo{} deletions in a
  region of chromosome 22 previously implicated in DiGeorge
  syndrome. These regions were validated \via{} multiplex qPCR.  This
  paper also describes a simple tool for detection of \denovo{} copy
  number variants in case-parent designs that appears to be more
  robust to genomic waves and noisy data than existing approaches for
  \denovo{} inference.

\end{abstract}

\section{Introduction}

Cleft palate and cleft lip affect 1 in every 2500 offspring. This is
the first paper to identify copy number variants implicated in
clefting (?)

\section{Methods}


BeadStudio was used to process the raw intensities for 7599 samples,
including 318 HapMap samples.

\section{Results}

What we still need:

- plots for every chromosome with color codes for the type of denovo event

- For the rectangle plots:
    - use the pages option.  put everything we need in a big matrix?

    - make use of darker colors without borders for color coding calls
      Perhaps 3 colors for denovo hemizygous, denovo homozygous,
      denovo amplification (3 colors)

   - For chromosome-level plots, add a histogram with frequencies
     for the samples

   - Add a legend for colors.




\section{Discussion}



\end{document}

\section{Session Info}

<<sessionInfo, echo=FALSE, results=tex>>=
toLatex(sessionInfo())
@

