<<>>=
library(GenomicFeatures)
library(rtracklayer)
session <- browserSession()
genome(session) <- "hg18"
geneTable <- makeTranscriptDbFromUCSC(genome = "hg18", tablename = "knownGene")

tx.dump <- as.list(geneTable)

query <- ucscTableQuery(session,
			GenomicRanges(289991, 296156, "chr6"))



    txdb_file <- system.file("extdata", "UCSC_knownGene_sample.sqlite",
                                package="GenomicFeatures")
       txdb <- loadFeatures(txdb_file)

       ## Get the transcripts grouped by gene:
       transcriptsBy(txdb, "gene")

       ## Get the exons grouped by gene:
       exonsBy(txdb, "gene")

       ## Get the cds grouped by transcript:
       cds_by_tx0 <- cdsBy(txdb, "tx")
       ## With more informative group names:
       cds_by_tx1 <- cdsBy(txdb, "tx", use.names=TRUE)
       ## Note that 'cds_by_tx1' can also be obtained with:
       names(cds_by_tx0) <- id2name(txdb, feature.type="tx")[names(cds_by_tx0)]
       stopifnot(identical(cds_by_tx0, cds_by_tx1))

       ## Get the introns grouped by transcript:
       intronsByTranscript(txdb)

       ## Get the 5' UTRs grouped by transcript:
       fiveUTRsByTranscript(txdb)
       fiveUTRsByTranscript(txdb, use.names=TRUE)  # more informative group names

@

<<>>=
library("biomaRt")
listMarts()
ensembl=useMart("ensembl")
listDatasets(ensembl)
ensembl = useDataset("hsapiens_gene_ensembl",mart=ensembl)
## alternatively
##ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
##filters = listFilters(ensembl)
##filters[1:5,]
##attributes = listAttributes(ensembl)
##attributes[1:5,]

##affyids=c("202763_at","209310_s_at","207500_at")
##getBM(attributes=c('affy_hg_u133_plus_2', 'hgnc_symbol', 'chromosome_name','start_position','end_position', 'band'),
## filters = 'affy_hg_u133_plus_2', values = affyids, mart = ensembl)

jugessur <- read.csv("../extdata/snp_names_in_jugessur.csv")
##chrBands <- as.character(jugessur[, 4])
##chrBands <- sapply(chrBands, function(x) strsplit(x, "-")[[1]][[1]])
##jugessur$chr.band <- chrBands

biomart.table <- getBM(attributes=c("hgnc_symbol", "band", "chromosome_name", "start_position", "end_position"),
		       values=chrBands, mart=ensembl)

##biomart.table <- biomart.table[biomart.table$chromosome %in% 1:22, ]
##biomart.table$chr.band <- paste(biomart.table$chromosome, biomart.table$band, sep="")

##tab <- merge(jugessur, biomart.table)
index <- match(unique(jugessur$gene), biomart.table$hgnc_symbol)
index <- index[!is.na(index)]
biomart.table <- biomart.table[index, ]
jugessur.ranges <- RangedData(IRanges(biomart.table$start_position,
				      biomart.table$end_position),
			      chrom=as.integer(biomart.table$chromosome_name),
			      hgnc_symbol=biomart.table$hgnc_symbol,
			      band=biomart.table$band)
save(jugessur.ranges, file="../../data/jugessur.ranges.rda")
@
