%\VignetteIndexEntry{Detection of de novo copy number alterations in case-parent trios}
%\VignetteDepends{?}
%\VignetteKeywords{MinimumDistance, copy number, SNP, case-parent trios, de novo}
%\VignettePackage{MinimumDistance}
\documentclass{article}
\usepackage{graphicx}
\usepackage{natbib}
\usepackage{url}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rcode}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textsf{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}
\newcommand{\R}{\textsf{R}}
\newcommand{\md}{\Rpackage{MinimumDistance}}
\newcommand{\md}{\Rpackage{MinimumDistance}}
\newcommand{\logRratio}{$\log_2$ R ratio}
\newcommand{\baf}{B allele frequency}
\newcommand{\bafs}{B allele frequencies}
\newcommand{\tsl}{trioSetList}
\newcommand{\ts}{trioSet object}
\usepackage[margin=1in]{geometry}


\begin{document}
\title{Detection of de novo copy number alterations in case-parent
  trios using the \R{} package \md{}}
\date{\today}

\author{Moiz Bootwalla and Rob Scharpf}
\maketitle


<<setup, echo=FALSE, results=hide>>=
options(prompt="R> ", continue=" ", device=pdf, width=68)
@

\begin{abstract}

  This vignette provides an example of computing the minimum distance
  using the \R{} package \md{} which is used to detect de novo copy
  number alterations in case-parent trios from HapMap Phase 2 Data.

\end{abstract}

\section{Exemplar dataset}

This section should describe what data is required for the package and
how to format the data appropriately for reproducing our analysis.
{\it Ideally, we would include processed data in the package so that
  users can reproduce all of the analytic steps after preprocessing
  (e.g., after the BeadStudio analysis). }

This package requires
\begin{itemize}
  \item a \tsl{}, each element of which is a \ts{}. Apart from the
    regular slots in an \Rclass{eSet} object a \ts{} has slots for
    \logRratio{}s, \bafs{} and minimum distance. Both the
    \logRratio{}s and \bafs{} are organized as three dimensional
    arrays; the dimensions of which are number of markers X number of
    trios X 3 (the members in a trio i.e. Father, Mother and Offspring)
  \item a pedigree file, each row of which contains the pedigree
    information of the samples that make up a trio.
  \item a samplesheet file which contains information on gender, DNA
    source, chemistry plate, etc. This file must contain a Sample.Name column.
  \item \ldots
\end{itemize}

An example of each of these files is provided with the \R{} package
\md{}. They can be accessed as follows:

<<exampleFile>>=
library(MinimumDistance)
data(pedigreeExample)
data(samplesheetExample)
data(trioSetList)
@

The sample sheet is in the same directory 'HumanHap610Quadv1_Sample_Map.csv'.

We really only need one father, mother, offspring trio for the
vignette, but it would be good to find an example of a trio with a de
novo amplification or deletion.  If not too big, we might include the
data for a few of the trios and a subset of the chromosomes (e.g., 5
trios and 4 chromosomes).  Example of the format currently required -- see

``/thumper/ctsa/beaty/holger/txtfiles''

You'll need the physical position and chromosome info.  The annotation
packages should be available from BioConductor. Here's how you can see
what annotation packages are available.  (You'll want to include this
in the vignette since other people may have different platforms).  The
R function minimumDistance initializes an object of class
'TrioSetList' and automatically provides annotation of chromosome and
physical position provided that the 'cdfname' for the package is
provided and an annotation package is available.

<<>>=
library(oligoClasses)
all.pkgs <- annotationPackages()
@

Show how to construct the container we use for analyzing trios from
the above files.

\begin{itemize}

  \item create matrix of B allele frequencies

  \item create matrix of \logRratio{}s

  \item create matrix of \bafs{}

  \item instantiate an object of class \Rclass{TrioSetList} object

  \end{itemize}

<<eval=FALSE>>=
container <- minimumDistance(path, ## for parsed data
			     samplesheet=samplesheet2,
			     pedigree=pedigree,
			     container.filename=container.filename,
			     chromosomes=chromosomes,
			     file.ext=".txt",
			     readFiles=readFiles,
			     calculate.mad=calculate.mad,
			     calculate.md=calculate.md,
			     exclusionRule=BeatyExperimentData:::excludeWGA,
			     cdfName="human610quadv1b")
@

<<>>=
load("~mbootwal/packages/MinimumDistance/data/trioSet-hapmap.rda")
##
library(MinimumDistance)
trioSetList <- minimumDistance(container=trioSetList, readFiles=FALSE, calculate.md=TRUE,
		       container.filename="~/trioSetList.rda")
##mads <- matrix(NA, ncol(trioSetList[[1]]), 3)
##colnames(mads) <- c("F", "M", "O")
##rownames(mads) <- sampleNames(trioSetList[[1]])
#### need to calculate mad
##fff <- function(object, trio.index){
##	lR <- logR(object)[, trio.index, ]
##}
any(is.na(mad(trioSetList[[1]])))
any(is.na(mad(trioSetList[[22]])))
##for(j in seq(length=length(rownames(mads)))){
##	logrs <- do.call("rbind", sapply(trioSetList, fff, trio.index=j))
##	mads[j, ] <- apply(logrs, 2, mad, na.rm=TRUE)
##}
##for(k in seq(length=length(trioSetList))){
##	mad(trioSetList[[k]]) <- mads
##}
##trace(MinimumDistance:::computeLoglik, browser)
unlink("~/results.rda")
rd <- minimumDistanceCalls(container=trioSetList, chromosomes=1:5, cbs.filename="~/results.rda")
@


\section{Detection of de novo copy number}

\subsection{Segmentation}

The function call looks something like this.

<<minimumDistanceCalls>>=
 ## segmentation
 ## classification
@


\subsection{Posterior probabilities of trio copy number states}

\subsection{Visualizations}

Plot the MAD log R ratio by chemistry plate

Plot the MAD log R ratio by DNA source

Plot the low-level summaries along with the minimum distance. Right
now, the function is called \Rfunction{minimumDistancePlot}.

'Rectangle' plots as in the paper (here, we'd only have de novo
predictions from 5 offspring)



\section{Session information}
<<sessionInfo, results=tex>>=
toLatex(sessionInfo())
@

\end{document}
