% \paragraph{Challenges}

There are several challenged to identifying denovo deletions in the cleft palate data, including batch effects and issues of DNA sample quality.  Our initial strategy for identifying denovo alterations will focus on the Beadstudio output (the log R ratios and B allele frequencies) for samples with high DNA quality, following by circular binary segmentation to identify regions in which the latent copy number is inferred to be constant.  Using simple rules that take into account the segment mean and the sample variance of the log R ratios, we will deletions independently for each individual.  We will classify a deletion event in the offspring as denovo only if there was no overlapping deletion in either parent.  Finally, we will prioritize our list of denovo events by the frequency at which the region was denovo across trios, the number of markers in the denovo region, and whether the region disrupts a coding region of the genome.  We expect that such an approach would be useful for generating a small list of 'top' scoring regions that can be validated by other platforms such as qPCR.  We will explore improvements to the algorithm that would allow allele-specific estimation of copy number and uncertainty and hidden markov models that infer denovo events directly.  We expect that these improvements could identify additional regions that are undetected in the initial algorithm, but could also reduce the number of false positives by using a likelihood based approach that takes into account uncertainty of the point estimates.

DNA quality.  A robust estimate of sample variance is the median absolute deviation (MAD) of the Beadstudio log R ratios taken across all autosomal chromosomes.  We expect that the MAD will reflect technological noise in the log R ratios.  In Figure X, we plot the MAD stratified by the DNA source and population ancestry. This plot shows that much of the technological noise in the Beadstudio estimates are explained by the DNA source.  In particular, samples that underwent whole genome amplification as part of the DNA preparation tend to have greater levels of technological variance than samples that were obtained from, for instance, blood or buccal swabs. The resolution to detect deletions and amplifications will be much lower in such samples.  As the number of deletion events detected by circular binary segmenation (CBS) in these samples was unusually high (data not shown), we have excluded these samples in our initial analysis to avoid a bias towards samples with high variance.  We expect that approaches that more directly model the uncertainty would lessen the need to filter these samples from our analysis.

Smoothing. We have used circular binary segmentation (CBS) to segment the log R ratios from the Beadstudio software (Olshen et al., 2004). CBS identifies genomic regions in which the inferred copy number is constant, but does not directly call regions as deleted or amplified.

Calling denovo deletions.  We have implemented simple rules to call deletions in parents and affected offspring that take into account both the segment mean and the MAD.  Specifically, we call an event as deleted in the offspring if the segment mean was less than the logarithm of 1/2 or -2 * the MAD, whichever was less.  For parents, we called a deletion when the segment mean was less than the logarithm of 1.5/2 or less than -1.5 * the MAD.  Note that this calling rule is more stringent for the offspring and more liberal for the parents, reducing the chance that an inferred denovo alteration is due to a false positive in the offpsring or a false negative in the parents. Finally, we call a deletion in the offspring as denovo if there was no overlapping region in the parents that was called a deletion.

Priotizing denovo regions. We will prioritize the list of denovo mutations by the frequency in which a denovo event occurs across the trios, whether the disruption occurs in a coding region of the genome, and the size of the inferred denovo alteration.



References:

Olshen AB, Venkatraman ES, Lucito R, and Wigler M (2004). Circular binary segmentation for the analysis of array-based DNA copy number data. Biostatistics 5 (4), 557-572.
